drivers/timer/cc13x2_cc26x2_rtc_timer.c

```diff
@@ -190,10 +190,10 @@ int z_clock_driver_init(struct device *device)
 	initDevice();
 	startDevice();
 
-	/* Enable RTC interrupt. */
-	IRQ_CONNECT(DT_INST_0_TI_CC13XX_CC26XX_RTC_IRQ_0,
-		DT_INST_0_TI_CC13XX_CC26XX_RTC_IRQ_0_PRIORITY,
-		rtc_isr, 0, 0);
+//	/* Enable RTC interrupt. */
+//	IRQ_CONNECT(DT_INST_0_TI_CC13XX_CC26XX_RTC_IRQ_0,
+//		DT_INST_0_TI_CC13XX_CC26XX_RTC_IRQ_0_PRIORITY,
+//		rtc_isr, 0, 0);
```

Comment:
It's hackishly called from subsys/bluetooth/controller/ll_sw/ti/lll/lll.c but definitely needs a better, permanent solution. Currently, shared IRQs are really not used in Zephyr. I had some discussion about it on the 'general' channel on slack. Shared IRQ's also do not work with Device Tree. So others have used a publish-subscribe pattern - so in `rtc_isr()` check the status of *all* channels, and then fire off callbacks with a parameter indicating which channel was active. The main isr would be responsible for clearing the event on the channel.

But then there is also the issue of keeping track of which channels are being used by which drivers, so it's a competition for shared resources. We can't have several different drivers overwriting the comparator value for a particular RTC channel. The obvious solution is to use a generic timer API that manages what timeouts occur on which channels on the RTC.

Of course, that is something that Zephyr has already had forever - I kind of wonder if it's possible to just use the general-purpose Zephyr timer to wake up the BT stack rather than have it manage the timer hardware directly. @cvinayak @carlescufi - I believe I saw a comment elsewhere where that idea was posed. The question was whether or not using the general-purpose timer would provide fast-enough and deterministic-enough response times for BLE.

Reply:
I'm going to defer fixing this until later, if that's ok, since getting a functional BLE implementation is my priority. The way that it works here, is that the RF driverlib owns the RTC IRQ and then it executes a callback in cc13x2_cc26x2_rtc_timer.c, but that's obviously kind of a hack.

@vanti, @bwitherspoon, it might make sense to have the cc13x2_cc26x2_rtc_timer.c also simply use the DPL via the [ClockP.h API](https://github.com/vanti/hal_ti/blob/master/simplelink/source/ti/drivers/dpl/ClockP.h). That's kind of the inverse of what is happening in upstream, but the benefit would be that it solves the shared interrupt problem in a common, scalable way. A config option could be provided to cc13x2_cc26x2_rtc_timer.c and the DPL to specify exactly which RTC channel they each operate on.

Also, while it might be best to dedicate RTC_CH1 to BLE and RTC_CH0 to the timer, it should be quite possible to *share* a single RTC channel (or any of the low-power timers) by organizing timer events using a min-heap data structure O(1) lookup first / remove or otherwise an ordered-map (a.k.a. hash-map, red-black tree, or AVL tree) O(NlogN). 

- [ ] Come to a consensus about the RTC interrupt

Reply:
Covered under the task list.

Reply:
@cfriedt Using the DPL would be possible but as you said, it would be contrary to what would be the usual goal of making Zephyr code more generic and less depending on vendor-specific code. I want to better understand the use-case here before we opt for a route. I now see rtc_isr is called in rtc0_cc13xx_cc26xx_isr (thanks for pointing that out). rtc0_cc13xx_cc26xx_isr is owned by the 'll_sw' layer in your current implementation. Ideally I think the 'll_sw' layer should  depend on the timer driver, and the timer driver should own the timer interrupt and its isr. "ll_sw" would register a callback with the timer driver that the latter would invoke. But the following limitations are preventing this flow:

1. There is no provision in the timer driver today to register callbacks that would be called directly by the isr
2. There is support for only one RTC channel in the timer driver

Does that sound right? If so, maybe we could look into solving these issues.

Your idea of sharing a single RTC channel sounds interesting. However, when I look at the code, I see RTC CH1 is being used by the counter module in the bluetooth controller, which provides an API to start/stop the RTC channel and set its comparator. If we were sharing RTC CH0 instead, which is used to keep track of the system clock, wouldn't this cause it to lose track of time? There must be something I am missing here.

Or perhaps that is what you meant by using "Zephyr's generic timer API": instead of having the counter module directly interface with the timer hardware, change it to use the [k_timer API](https://docs.zephyrproject.org/latest/reference/kernel/timing/timers.html) instead. k_timer would provide you with a 'timer instance' that would be based off of the system clock, effectively sharing RTC channel 0. Its expiry function, invoked in interrupt context, can be used to perform the work currently done in rtc0_cc13xx_cc26xx_isr (minus the call to rtc_isr and the code to deal with rtc ch1). In fact, ClockP is just a thin layer that ultimately calls into k_timer, so you should get equivalent functionality by using k_timer instead. There are potential concerns about latency, determinism and accuracy, etc., but short of better knowledge, I'd encourage you to try this approach, as it'd simplify the system by not involving a second RTC channel.

We can keep discussing this while you get the implementation working. I think it would be fine to defer fixing this issue for now.